rfcs/0006-transaction.md

```diff
@@ -0,0 +1,299 @@
+
+## Proposal
+
+### API
+
+The transaction API should be resembles with the non-transactioned API. All the api like `get()`, `put()`, `range()` should also be supported in the transaction API, and these APIs should be aligned with the Snapshot API + WriteBatch API. The only difference is that the transaction API should be able to `commit()` or `rollback()` the changes.
+
+(note: at the time of this RFC, the Snapshot API and WriteBatch API are still not implemented, the transaction feature should be implemented after these APIs are implemented.)
+
+The write operations like `put()` in the transaction should be buffered in the transaction state, and only be written to the underlying storage when `commit()` is called.
+
+```rust
+let object_store: Arc<dyn ObjectStore> = Arc::new(InMemory::new());
+let options = DbOptions::default();
+let kv_store = Db::open_with_opts(
+    Path::from("/tmp/test_kv_store"),
+    options,
+    object_store,
+)
+.await
+.unwrap();
+
+// start transaction and write some data
+let txn = kv_store.transaction();
+let val1 = txn.get("key1").await;
+
+// could batch the write operations
+txn.put("key2", val1);
+txn.put("key3", "val3");
+
+// should see the new value of key2
+assert_eq!(txn.get("key2").await, val1);
+
+// commit, might fail on conflict, when it fails, the transaction should be rolled back
+txn.commit().await?;
+```
+
+### Conflict Checking
+
+The approach we choose on conflict checking is expected to be more similar to Badger's global `Oracle` approach, since we hope to provide SSI in the transaction feature, and it's less prone to unnecessarily abort the transaction with limited MemTable size.
+
+Below is the data structure of the global `Oracle` and the transaction state:
+
+```rust
+struct Oracle {
+    next_seq: AtomicU64,
+    recent_committed_txns: Deque<Arc<TransactionState>>,
+}
+
+struct TransactionState {
+    started_seq: u64,
+    write_keys: HashSet<Bytes>,
+    read_keys: Vec<Range<Bytes>>,
+    committed_seq: Option<u64>
+}
+```
+
+`Oracle` is considered a global singleton in the DB, and it tracks the recent committed transactions in the `recent_committed_txns` deque. When a transaction commits, it should push itself into the `recent_committed_txns` deque, and the entries in `recent_committed_txns` can be GCed after they are no longer needed. Also, when a `TransactionState` is tracked in the `recent_committed_txns`, only the write keys need to be tracked, while `read_keys` should always be empty.
+
+Besides the started sequence number of the transaction, we also need to track the sequence number of the transaction committed. During the execution of a transaction, some other transactions might commit, and the current transaction should check the conflicts with these committed transactions.
+
+The conflict check simply checks whether the Read & Write keys of the current transaction are overlapped with the Write keys from related recent committed transactions.
```

Comment:
I've spent some time thinking about how the memtable implementation should look like with multi version keys, and there is one thing I keep returning to.

Is it correct that you not only need to check conflicts between transactions, but even against any non-transactional updates to the database?

For example:
```
        txn: <--------- (GET foo) (PUT bar) ---------- --------->
non-txn:                                                       (PUT foo)
```

It feels like during commit we always need to check all `read_keys` against memtable and make sure there was no override.

Is it correct? (sorry if this already covered somewhere, I might have missed it)

Reply:
>Is it correct that you not only need to check conflicts between transactions, but even against any non-transactional updates to the database?

Yep, correct! This is mentioned briefly here:

https://github.com/slatedb/slatedb/pull/260#discussion_r1855315369

Reply:
> Is it correct that you not only need to check conflicts between transactions, but even against any non-transactional updates to the database?

yes exactly, all the writes are considered as transactional, they're transactions with a single write.

Reply:
just curious, if there's no ongoing transactions can we skip this step for normal writes? probably premature optimization, but just curious to see if I got my head wrapped around this